gene/insert.go

```diff
@@ -0,0 +1,326 @@
+		frame := (codingPos + 1) % 3
+		var currCodon dna.Codon
+
+		if frame != 0 { // if insertion disrupts a codon, note the codon being changed
+			currCodon, err = CdnaPosToCodon(g, codingPos)
+			if err != nil { //TODO REMOVE
+				log.Panic("ERROR WITH CDNA POS")
+			}
+			answer.AaRef = []dna.AminoAcid{dna.TranslateCodon(currCodon)}
+		}
+
+		// Update cDNA
+		err = insertAdjust(g, &g.cdnaSeq, codingPos+1+len(g.utrFive.seq), alt)
+		if err != nil {
+			return EffectPrediction{}, err
+		}
+
+		var newProtSeq []dna.AminoAcid
+
+		answer.AaPos = codingPos / 3
+
+		if len(alt)%3 != 0 { // Causes Frameshift
+			answer.StopDist = 0
+			answer.Consequence = Frameshift
+
+			newProtSeq = frameshiftTranslate(g.codingSeq.seq[(codingPos+1)-frame:], g.utrThree.seq)
+
+			if newProtSeq[len(newProtSeq)-1] == dna.Stop {
+				answer.StopDist = len(newProtSeq) - 1
+			} else {
+				answer.StopDist = -2
+			}
+
+			var j int
+			for j = 0; newProtSeq[j] == g.protSeq[answer.AaPos]; j++ { // trim matching amino acids from frameshift
+				answer.AaPos++
+				if answer.StopDist != -2 {
+					answer.StopDist--
+				}
+				if answer.AaPos >= len(g.protSeq) {
+					break
+				}
+			}
+
+			answer.AaRef = []dna.AminoAcid{g.protSeq[answer.AaPos]}
+			answer.AaAlt = []dna.AminoAcid{newProtSeq[j]}
+
+			g.protSeq = newProtSeq //TODO append new prot seq to existing seq at point of insertion
+
+		} else { // In-Frame
+			newProtSeq = dna.TranslateToTer(g.codingSeq.seq) //TODO this can be much more efficient
+			answer.Consequence = InFrameInsertion
+			if frame != 0 { // Disrupts an existing codon
+				answer.AaAlt = dna.TranslateSeq(g.codingSeq.seq[(codingPos+1)-frame : (codingPos+1)+len(alt)+(3-frame)])
+				if answer.AaRef[0] == answer.AaAlt[0] {
+					answer.AaRef = nil
+					answer.AaAlt = answer.AaAlt[1:]
+					answer.AaPos++
+				}
+			} else { // Does not disrupt existing codons AND is inframe
+				answer.AaAlt = dna.TranslateSeq(alt)
+			}
+		}
+		g.protSeq = newProtSeq
+	} else { // NonCoding
+		// Update featureArray for noncoding
+		// figure out how the inserted bases should be labelled in the feature array. Priority: 5'UTR=3'Utr>Intron
+		fillVal := Feature(numbers.Min(int(g.featureArray[genomeIndexPos]), int(g.featureArray[genomeIndexPos+1])))
+
+		switch fillVal {
+		case UtrFive:
+			utrFiveOffset := 0
+
+			for i := 0; i <= genomeIndexPos; i++ {
+				if g.featureArray[i] == UtrFive {
+					utrFiveOffset++
+				}
+				if i > len(g.featureArray) { // case of no codingSeq
+					break
+				}
+			}
+
+			if utrFiveOffset > -1 { // Update 5' UTR
+				err = insertAdjust(g, &g.cdnaSeq, g.utrFive.start+utrFiveOffset, alt)
+				if err != nil {
+					return EffectPrediction{}, err
+				}
+			}
+
+		case UtrThree:
+			utrThreeOffset := 0
+
+			for i := genomeIndexPos; g.featureArray[i] < 0; i-- {
+				if g.featureArray[i] == UtrThree {
+					utrThreeOffset++
+				}
+				if i <= 0 { // case of no codingSeq
```

Comment:
The everything annotated as 3' UTR when there is no CDS?

Reply:
No, this is catching a corner case where the entire gene is annotated as a 3' UTR. It won't actually change the annotation at all, but it will let insertAdjust know that the insertion should be made at the start of the gene + the number of bases preceding the genomeIndexPos.